File: src/day-7/i.py
from typing import Optional


class Dir:
    def __init__(self, root: Optional['Dir'], name: str, size: Optional[int], dirs: Optional[list['Dir']]) -> None:
        self.__root = root
        self.__name = name
        self.__size = size
        self.__dirs = dirs

    def root(self) -> Optional['Dir']:
        return self.__root

    def name(self) -> str:
        return self.__name

    def size(self) -> Optional[int]:
        if self.__size is None:
            self.__size = self.__calc_size(self.__dirs)

        return self.__size

    def dirs(self) -> Optional[list['Dir']]:
        return self.__dirs

    def set_dirs(self, dirs: Optional[list['Dir']]):
        self.__dirs = dirs

    def __calc_size(self, dirs: Optional[list['Dir']]) -> int:
        if dirs is None:
            return self.__size
        elif self.__size is None:
            self.__size = 0

        return self.__size + sum(map(lambda dir: dir.size(), dirs))


def main():
    root = Dir(None, '/', None, [])
    current_dir = root
    current_dir_dirs = []

    with open('i.in', 'r') as f:
        for line in f.readlines():
            split = line.split(' ')
            first = split[0]
            match first:
                case '$':
                    if current_dir is not None:
                        current_dir.set_dirs(current_dir_dirs)
                    command = split[1]
                    match command:
                        case 'ls':
                            current_dir_dirs = []
                        case 'cd':
                            target_dir = split[2]
                            match target_dir:
                                case '/':
                                    pass
                                case '..':
                                    current_dir = current_dir.root()
                                case _:
                                    if current_dir is not None:
                                        current_dir = next(
                                            filter(lambda dir: dir.name() == target_dir, current_dir.dirs()), None)
                case 'dir':
                    dir = split[1]
                    current_dir_dirs.append(Dir(current_dir, dir, None, []))
                case size_str:
                    size = int(size_str)
                    file = split[1]
                    current_dir_dirs.append(Dir(current_dir, file, size, None))

    print(root.size())


main()

File: src/day-7/i.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::fs;
use std::rc::Rc;

mod util;
use util::*;

mod dir;
use dir::{Cwd, Dir};

fn main() {
    let root = Rc::new(RefCell::new(Dir::new(None, 0, HashMap::new())));
    let mut current_dir = root.clone();

    let mut total = 0;

    let _input = fs::read_to_string("i.in").unwrap();

    for line in _input.lines() {
        let split = line.split(" ");
        let split = split.collect::<Vec<&str>>();

        let mut i = 0;
        let first = *split.get(0).unwrap();
        match first {
            // command
            "$" => {
                let command = *split.get(1).unwrap();
                i += 1;
                if i == 2 {
                    break;
                }
                match command {
                    "ls" => {
                        // we can ignore
                    }
                    "cd" => {
                        let cd_target = *split.get(2).unwrap();
                        match cd_target {
                            ".." => {
                                let parent = current_dir.borrow().root.clone().unwrap();
                                current_dir = parent;
                            }
                            _ => {
                                let child = current_dir.borrow_mut().dirs.entry(cd_target).or_default().clone();
                                current_dir = child;
                            }
                        }
                    }
                    _ => unimplemented!(),
                }
            }
            // directory
            "dir" => {
                let dir_name = *split.get(1).unwrap();
                current_dir
                    .borrow_mut()
                    .add_dir(dir_name, Dir::new(None, 0, HashMap::new()));
            }
            // file
            _ => {
                let size = first.parse::<u32>().unwrap();
                let file_name = *split.get(1).unwrap();
                current_dir.borrow_mut().add_dir(file_name, Dir::new(Some(current_dir.clone()), size, HashMap::new()));
            }
        }
    }
    println!("{total}");
    println!("{current_dir:#?}");
}

// // fn update_dirs(root: &mut Dir, )
// use std::ref;

// use std::cell::RefCell;

// #[derive(Debug)]
// struct N {
//     pub root: Option<Box<RefCell<N>>>,
//     pub size: u32,
//     pub n: Option<Vec<N<>>>,
// }

// impl N {
//     fn update_n(&mut self, n: Option<Vec<N>>) {
//         self.n = n;
//     }
// }

// fn main() {
//     let mut n = N { root: None, size: 1, n: None };
//     n.update_n(None);
//     n.update_n(Some(vec![N { root: Some(RefCell::, size: 3, n: None }]));
//     dbg!(n);
// }
